src/qobj/arithmetic_and_attributes.jl

```diff
@@ -36,84 +36,81 @@ end
 
 for op in (:(+), :(-), :(*))
     @eval begin
-        function LinearAlgebra.$op(
-            A::QuantumObject{<:AbstractArray{T1},OpType},
-            B::QuantumObject{<:AbstractArray{T2},OpType},
-        ) where {T1,T2,OpType<:QuantumObjectType}
-            A.dims != B.dims &&
-                throw(DimensionMismatch("The two quantum objects are not of the same Hilbert dimension."))
+        function LinearAlgebra.$op(A::AbstractQuantumObject, B::AbstractQuantumObject)
+            check_dims(A, B)
+            if A isa QuantumObjectEvolution || B isa QuantumObjectEvolution
+                return QuantumObjectEvolution($(op)(A.data, B.data), A.type, A.dims)
+            end
             return QuantumObject($(op)(A.data, B.data), A.type, A.dims)
```

Comment:
I think we can extend the basic julia function `promote_type` (maybe put it in the file `utilities.jl` ?) for this.
something like
```julia
promote_type(A::QuantumObjectEvolution, B::QuantumObjectEvolution) = get_typename_wrapper(A)
promote_type(A::QuantumObjectEvolution, B::QuantumObject) = get_typename_wrapper(A)
promote_type(A::QuantumObject, B::QuantumObjectEvolution) = get_typename_wrapper(B)
promote_type(A::QuantumObject, B::QuantumObject) = get_typename_wrapper(A)
```

and therefore, this kind of `if`-condition block can be simplified as
```julia
QType = promote_type(A, B)
QType($(op)(A.data, B.data), A.type, A.dims)
```